# Importing Libraries

In [6]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [7]:
path = r'c:\users\vinor\01-2020 Instacart Basket Analysis'

In [8]:
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'),index_col = False)

In [9]:
df_ords = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'),index_col = False)

# 1.Data Consistency Checks in df_prods dataframe

In [10]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

it seems that there is an anomaly in the "prices" column, specifically in the maximum value. The maximum price is listed as 99,999, which is significantly higher than the other values in the dataset. This value is highly unusual and might indicate an error or outlier in the data.
To address this anomaly, you may want to investigate the source of the data or check for data entry errors

In [6]:
# Mixed-Type Data
# Create a dataframe
df_test = pd.DataFrame()

In [7]:
# Create a Mixed type column
df_test ['mix'] = ['a', 'b', 1, 'True']

In [8]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\vinor\AppData\Local\Temp\ipykernel_19220\2671172800.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [9]:
# convert a column’s data type from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
# Finding Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Create a new dataframe, df_nan, containing only those values within the "product_name" 
# column that meet the condition isnull() = True. 
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing Missing Values

In [13]:
# compare the number of rows in your current dataframe
df_prods.shape

(49693, 5)

In [14]:
#  non-missing values in new dataframe 
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
# to check that the number of rows has decreased
df_prods_clean.shape

(49677, 5)

# Duplicates

In [16]:
# command to look for full duplicates within the dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# Addressing Duplicates

In [18]:
# let’s check the current number of rows in the df_prods_clean dataframe
df_prods_clean.shape

(49677, 5)

In [19]:
# create a new dataframe that doesn’t include the duplicates 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

# 2. Data Consistency Checks in df_ords dataframe

In [21]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

I think in order_id column the max value is 3 but in 25% the value is 8 which is higher than max value, the same occurs in user_id column 25% is higher than max value and also in days_since_last_order column the max value is 3 but 50% value is 7 which is higher than max.

# 3. Check for mixed-type data in  df_ords dataframe.

In [22]:
# checking for mixed data types 
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\vinor\AppData\Local\Temp\ipykernel_19220\239215289.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\vinor\AppData\Local\Temp\ipykernel_19220\239215289.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\vinor\AppData\Local\Temp\ipykernel_19220\239215289.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\vinor\AppData\Local\Temp\ipykernel_19220\239215289.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\vinor\AppData\Local\Tem

In [29]:
# 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type
# There are no mixed data type in the df_ords dataframe

# 5. check for missing values in  df_ords dataframe.

In [25]:
df_ords.isnull().sum() 

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [27]:
# # Create a new dataframe, df_ords_nan, containing only those values within the "days_since_last_order" 
# column that meet the condition isnull() = True.
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [28]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

days_since_prior_order has 206,209 missing values. Generally, missing values are caused by either data corruption or not being recorded in first place. In this case, it's due to values not being recorded because days_since_prior_order only is null when the order is the first order from a user. In other words, the missing values indicate when order_number = 1

# 6. Address the missing values using an appropriate method.

In [33]:
# check whether order_number = 1 for days_since_prior_order = NaN to verify hypothesis that missing values
# indicate a user's first order
    
df_ords_prior_isnull = df_ords[df_ords['days_since_last_order'].isnull()==True]

In [34]:
# verify that order_number = 1 for all days_since_last_order = NaN

df_ords_prior_isnull['order_number'].describe()

count    206209.0
mean          1.0
std           0.0
min           1.0
25%           1.0
50%           1.0
75%           1.0
max           1.0
Name: order_number, dtype: float64

In [36]:
# create flag column for days_since_prior_order = Nan

df_ords['first_order'] = df_ords['days_since_last_order'].isnull() == True

In [37]:
# show df_ords

df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_last_order  first_order  
0                        8                    NaN         True  
1                        7                   15.0        False  
2                       12                   21.0        False  
3                        7                   29.0        False  
4                       15                   28.0        False  
...                    ...                    ...          ...  
3421078                 18                   29.0        False  
3421079                 10                   30.0        False  
3421080                 12                   18.0        False  
3421081                 12                    7.0        False  
3421082                 14                   30.0        False  

[3421083 rows x 8 columns]

In [40]:
#I chose to create a flag column because we know that the nulls are important, they indicate an user's first_order. 
#I could have chosen to not add a flag column since order_number = 1 indicates the same thing, 
#but I added it anyway for better readability.
#Imputing values would be incorrect and removing the observations would change the overall description of the data

# check for duplicate values

In [41]:
# check for full duplicates

df_ords_dups = df_ords[df_ords.duplicated()]

In [42]:
# show duplicates

df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order, first_order]
Index: []

In [44]:
# No duplicates found

In [45]:
# Export df_ords as orders_checked
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [46]:
# Export df_prods as products_checked
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))